src/access/AccessRestriction.h

```diff
@@ -0,0 +1,288 @@
+     *
+     * @param [in] fabricIndex  The index of the fabric for which to create entries.
+     */
+    CHIP_ERROR CreateFabricEntries(const FabricIndex fabricIndex);
+
+    /**
+     * Add a listener to be notified of changes in the access restriction list and active reviews.
+     *
+     * @param [in] listener  The listener to add.
+     */
+    void AddListener(EntryListener & listener);
+
+    /**
+     * Remove a listener from being notified of changes in the access restriction list and active reviews.
+     *
+     * @param [in] listener  The listener to remove.
+     */
+    void RemoveListener(EntryListener & listener);
+
+    /**
+     * Check whether access by a subject descriptor to a request path should be restricted (denied) for the given action.
+     * These restrictions are are only a part of overall access evaluation.
+     *
+     * If access is not restricted, CHIP_NO_ERROR will be returned.
+     *
+     * @retval #CHIP_ERROR_ACCESS_DENIED if access is denied.
+     * @retval other errors should also be treated as restricted/denied.
+     * @retval #CHIP_NO_ERROR if access is not restricted/denied.
+     */
+    CHIP_ERROR Check(const SubjectDescriptor & subjectDescriptor, const RequestPath & requestPath);
+
+    /**
+     * Request a review of the access restrictions for a fabric.
+     *
+     * @param [in]  fabricIndex  The index of the fabric requesting a review.
+     * @param [in]  arl          An optinal list of access restriction entries to review.  If null, all entries will be reviewed.
+     * @param [out] token        The unique token for the review, which can be matched to a review update event.
+     */
+    CHIP_ERROR RequestFabricRestrictionReview(FabricIndex fabricIndex, const std::vector<Entry> & arl, uint64_t & token)
+    {
+        token = ++mNextToken;
+        return DoRequestFabricRestrictionReview(fabricIndex, token, arl);
+    }
+
+    using EntryIterator = std::vector<SharedPtr<Entry>>::iterator;
+
+    /**
+     * Get iterator over the commissioning entries, which are used during commissioning
+     * and also installed as defaults for a fabric upon completion of commissioning.
+     *
+     * @param [out] begin iterator pointing to the beginning of the entries
+     * @param [out] end iterator pointing to the end of the entries
+     */
+    static CHIP_ERROR CommissioningEntries(EntryIterator & begin, EntryIterator & end)
+    {
+        begin = sCommissioningEntries.begin();
+        end   = sCommissioningEntries.end();
+
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * Get iterator over entries in the access restriction list by fabric.
+     *
+     * @param [in]  fabricIndex Iteration is confined to fabric
+     * @param [out] begin iterator pointing to the beginning of the entries
+     * @param [out] end iterator pointing to the end of the entries
+     */
+    CHIP_ERROR Entries(const FabricIndex fabricIndex, EntryIterator & begin, EntryIterator & end)
```

Comment:
Does this properly handle delete calls while holding the iterator?

Reply:
This data management model has been greatly simplified and this isnt an issue anymore.